### PR TITLE
feat(volume group): add in memory operations for volume group

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
1. Make SpecOperationsHelper trait optional to for resources.
2. Add in memory operations for volume group.
3. Intergate volume group with create volume.

### CODE REVIEW

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try-

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 1]
```diff
@@ -1115,13 +1141,24 @@ impl ResourceSpecsLocked {
...
 
-    /// Get or Create the resourced VolumeSpec for the given request
+    /// Remove volume group by its `id` only if the volume list becomes empty.
+    pub(super) fn remove_volume_group(&self, id: &VolumeId, vg_id: &String) {
+        let mut specs = self.write();
+        if let Some(vg_spec) = specs.volume_groups.get(vg_id) {
+            vg_spec.lock().remove(id);
+            if vg_spec.lock().is_empty() {
+                specs.volume_groups.remove(vg_id);
+            }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/specs.rs]: is this racy? After checking `is_empty` another one might be added? Maybe keep vg_spec locked while we remove it?

**Review by tiagolobocastro — APPROVED** [review]
We're not using the vg in the other scenarios where we add replicas, will that be on another PR?

File: control-plane/agents/src/bin/core/volume/specs.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/specs.rs]: > We're not using the vg in the other scenarios where we add replicas, will that be on another PR?

Yes, I will add the other operations, in the upcoming PRs. Till this point was mainly the structural part without any impact.

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**